### DISCUSSION TITLE
Alternate Packages

### DISCUSSION THREAD

**Discussion by owl4ce** [dslm4515/CMLFS#30] [Ideas]
Hi, I'm sharing the references I use. 

Here are the results I got:

## Successful with notes
- [x] `Ninja` -> `Samurai`
- [x] `Zlib` -> `Zlib-ng`
- [x] `Gzip` -> `Pigz`
- [x] `Pkg-config` -> `Pkgconf`
- [x] `GNU Gettext` -> `Gettext-tiny`
- [x] `GNU Ncurses` -> `NetBSD Curses`

> * **Gettext-tiny**: [notes by @dslm4515](https://github.com/dslm4515/CMLFS/issues/16#issuecomment-874946958)
> * **NetBSD Curses:** `Util-linux` fails to build, need to patch `more.c`
> 
>   ```sh
>   sed -i 's|, 0, 0|, 0, 0, 0, 0, 0, 0, 0, 0, 0|' text-utils/more.c
>   ```

## Failed with reasons
- [ ] `GNU M4`  -> `OpenBSD M4` or `Quasar M4`
> Caused by `GNU autoconf` (autom4te)
- [ ] `GNU Readline` -> `NetBSD libedit`  
> Caused by `Bash`
- [ ] `GNU Bison` -> `OpenBSD Yacc`  or `Byacc`
> Caused by `Util-linux` (autogen.sh)

...

**Comment by dslm4515** [comment]
Nice! 

I suppose for when building perl, you used the bundled zlib source included in the perl source, instead of building perl against system installed zlib-ng?

**Reply by owl4ce** [reply]
yeah, temporary.

**Reply by dslm4515** [reply]
Thought you might have gotten lucky like finding that patch for libelf!

**Comment by owl4ce** [comment]
I have notes, `libuv` for Cmake and `libffi` for Python are not required in `llvmtools` because it has those built-in libraries.

**Reply by dslm4515** [reply]
YES.
I forgot about that, when I realized i didnt need to add libarchive, curl, and the other deps for CMake. _I'll remove libuv for llvmtools._

But **libcxxabi requires python** during configuring... unless I missed a CMake option to not use python when building libcxxabi?

**Reply by owl4ce** [reply]
As far as I know, LLVM packages always required python. Idk, I've never tried without python, I'm just referencing the build packages of existing distros.

**Reply by owl4ce** [reply]
<br>

![image](https://user-images.githubusercontent.com/53987136/125996237-b0cd1e37-dd78-4af7-a5be-a08e6d39b3e1.png)


https://llvm.org/docs/GettingStarted.html#software

**Comment by owl4ce** [comment]
Found good references.
https://wiki.musl-libc.org/alternatives.html

**Reply by dslm4515** [reply]
I'm interested in...

- slibtool for ... libtool?
- cDetect for ... autotools? (autoconf, automake, etc)
- libsep... i think i use musl-rpmatch instead, i think?

**Reply by owl4ce** [reply]
I stick to GNU autotools, and mostly like in this repo.

**Reply by owl4ce** [reply]
May I ask, whats packages that required `musl-rpmatch`?

**Reply by owl4ce** [reply]
[cDetect](https://github.com/xyproto/cdetect) seems outdated.

**Reply by dslm4515** [reply]
> May I ask, whats packages that required `musl-rpmatch`?

Lol, that is a good question! When I found the musl-fts repo (i forgot why i needed FTS for musl) I also found the musl-rpmatch repo... thought I'd add it to MLFS/CMLFS just like musl-FTS.

I have built MLFS a few times and never needed musl-rpmatch... perhaps it should be marked as optional until a package (in BLFS/BMLFS) needs it?

**Reply by owl4ce** [reply]
> > May I ask, whats packages that required `musl-rpmatch`?
> 
> Lol, that is a good question! When I found the musl-fts repo (i forgot why i needed FTS for musl) I also found the musl-rpmatch repo... thought I'd add it to MLFS/CMLFS just like musl-FTS.
> 
> I have built MLFS a few times and never needed musl-rpmatch... perhaps it should be marked as optional until a package (in BLFS/BMLFS) needs it?

`musl-fts`, `musl-obstack`, and `argp-standalone` needed by `libelf`! except for `musl-rpmatch` 😄 

https://github.com/void-linux/void-packages/blob/master/srcpkgs/elfutils/template#L22

**Reply by dslm4515** [reply]
> musl-fts, musl-obstack, and argp-standalone needed by libelf!

YES! During early development of MLFS, libelf wont compile... realized i needed those 3 packages. It took me almost a year to figure out how to make a tool chain change libc's... GCC documentation wasnt so helpful, always assuming Glibc. Almost killed the project at the time.

**Reply by owl4ce** [reply]
> > musl-fts, musl-obstack, and argp-standalone needed by libelf!
> 
> YES! During early development of MLFS, libelf wont compile... realized i needed those 3 packages. It took me almost a year to figure out how to make a tool chain change libc's... GCC documentation wasnt so helpful, always assuming Glibc

Respectful.

**Reply by sskras** [reply]
@dslm4515 commented on [Jul 16, 2021](#discussioncomment-1014643):

> I'm interested in...
> 
> * slibtool for ... libtool?

Just in case... slibtool had 4 [new releases](https://github.com/midipix-project/slibtool/tags) since:

> [**v0.6.1**](https://github.com/midipix-project/slibtool/releases/tag/v0.6.1) on Dec 30, 2024
> [**v0.6.0**](https://github.com/midipix-project/slibtool/releases/tag/v0.6.0) on Jun 4, 2024
> [**v0.5.36**](https://github.com/midipix-project/slibtool/releases/tag/v0.5.36) on Mar 15, 2024
> [**v0.5.35**](https://github.com/midipix-project/slibtool/releases/tag/v0.5.35) on Nov 15, 2023

**Reply by takusuman** [reply]
Seems great.

**Reply by lm8** [reply]
I started that wiki page for alternatives.  Maintained it for a while too until I no longer had access.  I have lots of alternative packages I've found and patched or, in some cases, I've written from scratch.

**Comment by takusuman** [comment]
I don't mean to import Copacabana's dogfood into CMLFS --- more than I have already done since, I don't know, 2021 ---, but would you guys try ``opatch`` as the ``patch``(1) implementation?
https://github.com/Projeto-Pindorama/opatch
It is a port of OpenBSD's patch that I've made, I wish someone would use it besides me at Copacabana because, well, I'm preoccupied that the portability library I've written may have some bug that I couldn't find yet.

**Comment by andzai1995** [comment]
Gettext-tiny is a failed replacement of full GNU Gettext.

GRUB causes many errors when it was built with gettext-tiny due to lack of important files and successfully builds only with fully-featured GNU Gettext.

**Comment by andzai1995** [comment]
On 21st of June, 2025 released an alternative for Xorg Server - Xlibre project.
In Artix Linux it's available for testing in galaxy-gremlins repo.
https://github.com/X11Libre/xserver

**Reply by andzai1995** [reply]
XLibre XServer 25.0.0.4 running on CMLFS

![1751916144318](https://github.com/user-attachments/assets/0efad097-3cc8-423a-ae36-088ca7d474a4)
![1751916144281](https://github.com/user-attachments/assets/e18abb74-7042-4e8f-90fb-2ae6617105d6)
![1751916144354](https://github.com/user-attachments/assets/98cf1788-dfa5-44f1-9279-7b0e5abc767e)
